pkgs/build-support/build-incremental.nix

```diff
@@ -0,0 +1,47 @@
+{ pkgs }:
+rec {
+  /* Prepare a derivation for local builds.
+    *
+    * This function adds an additional output for a derivation,
+    * containing the build output.
+    * The build output can be used later to allow incremental builds
+    * by passing the `buildOut` output to the `mkIncrementalBuild` function.
+    *
+    * To build a project incrementaly follow these steps:
+    * - run prepareIncrementalBuild on the desired derivation
+    *   e.G `incrementalBuildArtifacts = (pkgs.buildIncremental.prepareIncrementalBuild pkgs.virtualbox).incrementalBuildArtifacts;`
+    * - change something you want in the sources of the package( e.G using source override)
+    *   changedVBox = pkgs.virtuabox.overrideAttrs (old: {
+    *      src = path/to/vbox/sources;
+    *   }
+    * - use `mkIncrementalBuild changedVBox buildOutput`
+    * - enjoy shorter build times
+  */
+  prepareIncrementalBuild = drv: drv.overrideAttrs (old: {
+    outputs = (old.outputs or [ "out" ]) ++ [ "incrementalBuildArtifacts" ];
+    installPhase = pkgs.lib.optionalString (!(builtins.hasAttr "outputs" old)) ''
+      mkdir -p $out
+    '' + (old.installPhase or "") + ''
+      mkdir -p $incrementalBuildArtifacts
+      cp -r ./* $incrementalBuildArtifacts/
+    '';
```

Comment:
Replacing the `installPhase` like this will break derivations that have no explicit `installPhase`, but `preInstall` or `postInstall` hooks. See https://nixos.org/manual/nixpkgs/stable/#sec-stdenv-phases:

> When overriding a phase, for example installPhase, it is important to start with runHook preInstall and end it with runHook postInstall, otherwise preInstall and postInstall will not be run. Even if you don’t use them directly, it is good practice to do so anyways for downstream users who would want to add a postInstall by overriding your derivation.

Would it make sense to copy the build artifacts in the `pre-` or `postInstall` hook instead, to avoid this pitfall?

Reply:
In fact, this PR could be implemented by registering a `pre`- or `postInstall` hook instead of doing this.

@messemar Did you consider cutting off the entire build at the install phase, and having only the incremental build artifacts as an output? This would avoid problems caused by a non-idempotent original `installPhase`, such as `mv foo $out/`. It also wastes less space, because the original outputs aren't actually used anyway.

Reply:
But yes, as its for development use only we can have a cut off after the install phase too.